### PR TITLE
[InternalImplementationOnly] support in the compilers

### LINKED ISSUE #220: Add support for InternalImplementationOnly attribute
InternalImplementationOnly attribute
====================================

One difficulty that faces API authors is that sometimes an `interface` is the most appropriate type for an API element (because there are multiple unrelated implementations internally). However, once a public API element is exposed as an `interface` it would be a breaking change to add additional methods in a future revision of the interface (clients who implement the interface would be broken because they do not implement the new members).

This issue is mitigated when the API is intended to be implemented only by its authoring assemblies. However, there is no language mechanism to enforce that clients do not implement the interface. API authors resort to documentation suggesting that the interface not be implemented in customer code, but customers can fail to notice the documentation resulting in painful compatibility constraints that are unintended.

The Roslyn APIs have an example of this issue with the [`ISymbol`](http://source.roslyn.io/#Microsoft.CodeAnalysis/Symbols/ISymbol.cs) interface and all of its derived interfaces such as [`IMethodSymbol`](http://source.roslyn.io/#Microsoft.CodeAnalysis/Symbols/IMethodSymbol.cs) and [`INamedTypeSymbol`](http://source.roslyn.io/#Microsoft.CodeAnalysis/Symbols/INamedTypeSymbol.cs). We intend to evolve these interfaces in the future as the language changes, but we may be prevented from doing so because customer code may extend these interfaces.

To mitigate this compatibility constraint we propose to add a compiler-provided error. Interfaces such as `ISymbol` that should only be extended or implemented in assemblies that have internal access are marked with an attribute of the name [`System.Runtime.CompilerServices.InternalImplementationOnlyAttribute`](http://source.roslyn.io/#Microsoft.CodeAnalysis/InternalImplementationOnlyAttribute.cs).

```cs
namespace System.Runtime.CompilerServices
{
    /// <summary>
    /// Indicates that the interface may only appear in the base clause of a type
    /// in the same assembly as the attributed interface or a type in
    /// an assembly that has InternalsVisibleTo the attributed interface's
    /// assembly.
    /// </summary>
    [AttributeUsage(AttributeTargets.Interface, AllowMultiple = false)]
    internal sealed class InternalImplementationOnlyAttribute : Attribute
    {
    }
}
```

This interface may be provided as a public type in a platform assembly, or as an internal type (as above) when the referenced platform does not yet provide direct support for the attribute.

When a compiler sees that such an attributed interface is extended or implemented outside of an assembly that has internal access, a compile-time error is produced.

```cs
namespace Microsoft.CodeAnalysis
{
    /// <summary>
    /// Represents a symbol (namespace, class, method, parameter, etc.)
    /// exposed by the compiler.
    /// </summary>
    [InternalImplementationOnly]
    public interface ISymbol : IEquatable<ISymbol>
    {
```

By adding this attribute to interfaces such as `ISymbol`, an API author can have some confidence that future additions to the interface will not break compatibility with existing clients, as those clients will have been prevented from implementing the interface.

### PR DESCRIPTION
**Pull request by gafter**
This fixes #220, moving the implementation of the InternalImplementationOnlyAttribute from an external analyzer (that can be turned off) to the compilers. F# to be done separately.

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet-bot test this please

**Comment by gafter** [comment]
@jaredpar How do I get this pull request to trigger a build-and-test run?

**Comment by gafter** [comment]
@VSadov @AlekseyTs Can you review this please?

**Comment by amcasey** [comment]
@gafter, I think you say @Jenkins test this please

**Comment by gafter** [comment]
@Jenkins test this please

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs [hunk 1]
```diff
@@ -131,11 +131,49 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
...
-                foreach (var @interface in interfaces)
+                foreach (var baseInterface in interfaces)
                 {
-                    @interface.CheckAllConstraints(conversions, location, diagnostics);
+                    baseInterface.CheckAllConstraints(conversions, location, diagnostics);
+                }
+
+                // we check direct base interfaces that have the InternalImplementationOnly attribute,
+                // giving an error when such an attributed interface appears in the interface list.
+                foreach (var baseInterface in this.InterfacesNoUseSiteDiagnostics(null))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: In general, there is no guarantee that bases of direct interfaces don't have the attribute.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs [hunk 2]
```diff
@@ -131,11 +131,49 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
...
+
+        /// <summary>
+        /// Return true of the given interface type contains an attribute of the
+        /// name System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private static bool HasInternalImplementationOnlyAttribute(NamedTypeSymbol baseInterface)
+        {
+            foreach (var att in baseInterface.GetAttributes())
+            {
+                if (att.IsTargetAttribute("System.Runtime.CompilerServices", "InternalImplementationOnlyAttribute"))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: This is a very inefficient way of checking for presence of a well-known attribute. In compiler, we should avoid realizing attributes on metadata and retargeting symbols. Source symbols can benefit from caching of the value too. I suggest doing the following:
- Follow the pattern of NamedTypeSymbol.HasSpecialName property. It is declared abstract and overridden as appropriate by derived classes. 
- Add attribute description to AttributeDescription. Detect and cache presence of the attribute on source symbols in SourceNamedTypeSymbol.DecodeWellKnownAttribute, follow SpecialNameAttribute pattern.
- For PENamedTypeSymbol, delegate check to PEModule, follow the pattern of PEModule.HasRequiredAttributeAttribute. Consider caching the result on the symbol if it can be done without significantly bloating it.

**Comment by AlekseyTs** [comment]
Consider separating feature implementation and attribute application to declared interfaces.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: Correct, and the tests ensure that it is allowed to indirectly implement or extend such an interface.

**Comment by gafter** [comment]
@AlekseyTs Can you please elaborate on what you mean by "Consider separating feature implementation and attribute application to declared interfaces"?

**Comment by AlekseyTs** [comment]
RE: Consider separating feature implementation and attribute application to declared interfaces.
I mean consider making changes like in IAliasSymbol.cs in a separate PR.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 3]
```diff
@@ -1053,7 +1053,7 @@ internal override void AddSynthesizedAttributes(ModuleCompilationState compilati
 
             if (this.Indexers.Any())
             {
-                string defaultMemberName = this.Indexers.First().MetadataName; // UNDONE: IndexerNameAttribute
+                string defaultMemberName = this.Indexers.First().MetadataName; // TODO: IndexerNameAttribute
                 var defaultMemberNameConstant = new TypedConstant(compilation.GetSpecialType(SpecialType.System_String), TypedConstantKind.Primitive, defaultMemberName);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: Overall I agree with moving UNDONE -> TODO but could we do it in a different commit?  It's making this change unnecessarily bigger.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols
+{
+    class InternalImplementationOnlyTests : CSharpTestBase
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs]: Please add explicit modifier

**Comment by jaredpar** [comment]
I agree with @AlekseyTs 

I think this change would be a lot cleaner if it were broken up into two parts:

1. Add the language feature to the compiler.
2. Change the compiler to use the language feature.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: Working on it.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: The approach you suggest would have the following performance impact:

1. Every attribute in source on a type would now have an additional check for this special attribute (in the sequence of in-then-else's in SourceNamedTypeSymbol.DecodeWellKnownAttribute) as well as an additional field (indirectly through the decoded well known attribute data) whenever the type is checked for any well known attribute. I believe that the situations where it is needed (inheriting from an interface in another assembly that has an annotation) is more rare than the situations where we would pay these costs (every annotation on every type). So there is a time and space penalty compared to the implemented approach (for source types). I'd prefer to use the current approach for source types.
2. For PE symbols, I like your approach but without caching since, again, the situation where the attribute is needed is rare.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs]: I can't tell where you're suggesting that modifier.

**Comment by gafter** [comment]
Unfortunately there are already uses of this attribute in the compiler code base, and those uses are not correct per the specification implemented here.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs (hunk 4 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs]: `internal class`

**Comment by jaredpar** [comment]
@gafter can' we then make it three changes

1. Rename the existing attribute to a new name.
2. Add the feature to the compiler.
3. Change the compiler to use the language feature.

Doing 2 and 3 and the same time is making it harder to tease apart the implementation and consumption.

**Comment by gafter** [comment]
I would rather not leave things in the intermediate state where there is nothing diagnosing the problem this issue is trying to address. Part of the point of mixing them together is that the consumption pattern has changed slightly, and this should the implementation of that change along with the consumption of that change.

**Comment by jaredpar** [comment]
If we don't want an i between state in master then the we should do this in stages in a side branch.

JaredPar from a phone
http://blog.paranoidcoding.com/
________________________________

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs]: Now I understand. (there is a modifier keyword `explicit` too)

**Comment by theoy** [comment]
Let's be careful about this pull request. I don't think we've closed this design/feature between all the stakeholders on whether this is the right direction for C#6/VB14 given the breaking changes and other consequences.

/cc @MadsTorgersen @MattGertz

**Comment by gafter** [comment]
I will do it in stages like this

1. Implement the compiler feature
2. Annotate the interfaces
3. Remove the analyzer

Then there is no gap, just a brief double-coverage.

**Comment by gafter** [comment]
@theoy Agreed, this is still under design review. I also agree that if this design is somehow rejected we'll need another mechanism to moderate the impact of the known planned future breaking changes to these interfaces, and we haven't figured out another mechanism that would be suitable. Whatever we do, it has to be soon to get into Dev14 and therefore protect us against future breaks.

**Comment by gafter** [comment]
OK, this pull request now contains only the language implementations for the attribute. Other changes will be done separately.

**Comment by gafter** [comment]
@Jenkins test this please

**Comment by gafter** [comment]
@theoy I just heard from @terrajobst they they hope to get to their review of this around Tuesday. If the approach is affirmed we can close on any remaining language issues on Wednesday.

**Comment by gafter** [comment]
@Jenkins test this please

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 5]
```diff
@@ -821,6 +821,22 @@ internal override bool HasSpecialName
             }
         }
 
+        internal override bool HasInternalImplementationOnlyAttribute
+        {
+            get
+            {
+                foreach (var att in this.GetAttributes())
+                {
+                    if (att.IsTargetAttribute("System.Runtime.CompilerServices", "InternalImplementationOnlyAttribute"))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: We should use helper that takes AttributeDescription.InternalImplementationOnlyAttribute so that shapes of constructors are checked and to be consistent with how we check for presence of this attribute in metadata.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 6]
```diff
@@ -821,6 +821,22 @@ internal override bool HasSpecialName
             }
         }
 
+        internal override bool HasInternalImplementationOnlyAttribute
+        {
+            get
+            {
+                foreach (var att in this.GetAttributes())
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: Please add a comment about why we think it is better to repeatedly iterate over attributes over an over again vs. caching the fact of the attribute presence similar to the other well-known attributes. Why it is a right thing to make an exception for this attribute.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs [hunk 7]
```diff
@@ -135,6 +135,27 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
...
+
+        /// <summary>
+        /// Produce an error if the given base interface may not be extended by this type due to the
+        /// presence of the attribute System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private void EnsureExtensible(NamedTypeSymbol baseInterface, SourceLocation location, DiagnosticBag diagnostics)
+        {
+            if (this.ContainingAssembly != baseInterface.ContainingAssembly &&
+                baseInterface.HasInternalImplementationOnlyAttribute &&
+                !this.ContainingAssembly.HasInternalAccessTo(baseInterface.ContainingAssembly))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: Would it make sense to perform HasInternalAccessTo before checking the attributes? Which check is cheaper?

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs [hunk 8]
```diff
@@ -185,6 +185,8 @@ internal override bool HasSpecialName
                 get { return false; }
             }
 
+            internal override bool HasInternalImplementationOnlyAttribute => false;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs]: Sealed?

File: src/Compilers/CSharp/Portable/Symbols/ErrorTypeSymbol.cs [hunk 9]
```diff
@@ -394,6 +394,8 @@ internal sealed override bool HasSpecialName
             get { return false; }
         }
 
+        internal override bool HasInternalImplementationOnlyAttribute => false;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ErrorTypeSymbol.cs]: Sealed?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs [hunk 10]
```diff
@@ -1835,6 +1835,9 @@ protected virtual DiagnosticInfo GetUseSiteDiagnosticImpl()
             return diagnostic;
         }
 
+        internal override bool HasInternalImplementationOnlyAttribute
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs]: Sealed?

File: src/Compilers/CSharp/Portable/Symbols/SubstitutedNamedTypeSymbol.cs [hunk 11]
```diff
@@ -145,6 +145,9 @@ internal sealed override bool HasSpecialName
             get { return _originalDefinition.HasSpecialName; }
         }
 
+        internal override bool HasInternalImplementationOnlyAttribute
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/SubstitutedNamedTypeSymbol.cs]: Sealed?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: I'd like to remove the constraint on the shape of the constructor, as we do not expect that to be part of the feature specification. How could we relax the check from metadata?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs [hunk 12]
```diff
@@ -306,5 +306,7 @@ internal override bool HasSpecialName
         {
             get { return false; }
         }
+
+        internal override bool HasInternalImplementationOnlyAttribute => false;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: Sealed?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 6 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: Will do.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs (hunk 7 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs]: I placed the tests in the order that I expected would minimize the total execution time of this code in a typical compilation.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    @"[assembly: System.Runtime.CompilerServices.InternalsVisibleTo(""C"")]
+public interface IB1 : IA1 {} // ok: has internal access
+public interface IB2 : IA2 {} // ok: IA2 not attributed
+public class B1 : IA1 {} // ok: has internal access
+public class B2 : IA2 {} // ok: IA2 not attributed";
+            var compb = CreateCompilation(
+                assemblyName: "B",
+                sources: new[] { bSource },
+                options: TestOptions.ReleaseDll,
+                references: new[] { new CSharpCompilationReference(compa), MetadataReference.CreateFromAssembly(typeof(System.Runtime.CompilerServices.InternalsVisibleToAttribute).Assembly) });
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs]: Should also test with metadata compa.EmitToImageReference()

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+public interface IC4 : IB2 {} // ok: IB2 not attributed
+public interface IC5 : IB1, IA1 {} // ** error: do not have internal access
+public interface IC6 : IB2, IA2 {} // ok: IA2 not attributed
+public class C1 : B1 {} // ok: B1 not attributed
+public class C2 : B2 {} // ok: B2 not attributed";
+            var compc = CreateCompilation(
+                assemblyName: "C",
+                sources: new[] { cSource },
+                options: TestOptions.ReleaseDll,
+                references: new[] { new CSharpCompilationReference(compa), new CSharpCompilationReference(compb), MetadataReference.CreateFromAssembly(typeof(System.Runtime.CompilerServices.InternalsVisibleToAttribute).Assembly) });
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/InternalImplementationOnlyTests.cs]: Should also test with metadata

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: I believe the shape of the constructor should be a part of the contract and we should enforce it as we do for any other well-known attribute.

File: src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/SynthesizedContainer.vb [hunk 15]
```diff
@@ -85,6 +85,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/SynthesizedContainer.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousTypeOrDelegatePublicSymbol.vb [hunk 16]
```diff
@@ -39,6 +39,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 End Get
             End Property
 
+            Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousTypeOrDelegatePublicSymbol.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousTypeOrDelegateTemplateSymbol.vb [hunk 17]
```diff
@@ -93,6 +93,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 End Get
             End Property
 
+            Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousTypeOrDelegateTemplateSymbol.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/ErrorTypeSymbol.vb [hunk 18]
```diff
@@ -165,6 +165,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/ErrorTypeSymbol.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb [hunk 19]
```diff
@@ -223,6 +223,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb [hunk 20]
```diff
@@ -2330,6 +2339,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
+            Get
+                For Each att In Me.GetAttributes
+                    If att.IsTargetAttribute("System.Runtime.CompilerServices", "InternalImplementationOnlyAttribute") Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb]: Use a helper that takes AttributeDescription

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb [hunk 21]
```diff
@@ -2330,6 +2339,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
+            Get
+                For Each att In Me.GetAttributes
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb]: Consider adding a comment mentioned for C# implementation.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol_ComClass.vb [hunk 22]
```diff
@@ -905,6 +905,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                     End Get
                 End Property
 
+                Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol_ComClass.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb [hunk 23]
```diff
@@ -60,6 +60,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Portable/Symbols/UnboundGenericType.vb [hunk 24]
```diff
@@ -67,6 +67,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        Friend Overrides ReadOnly Property HasInteralImplementationOnlyAttribute As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/UnboundGenericType.vb]: NotOverridable?

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+"
+            Dim compa = CreateVisualBasicCompilation(aSource, assemblyName:="A", referencedAssemblies:={ref})
+            compa.VerifyDiagnostics()
+            Const bSource = "<Assembly: System.Runtime.CompilerServices.InternalsVisibleTo(""C"")>
+Public Interface IB1 : Inherits IA1 : End Interface
+Public Interface IB2 : Inherits IA2 : End Interface
+Public Class B1 : Implements IA1 : End Class
+Public Class B2 : Implements IA2 : End Class
+"
+            Dim compb = CreateVisualBasicCompilation(code:=bSource, assemblyName:="B", referencedAssemblies:={ref}, referencedCompilations:={compa})
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb]: Should also test with metadata reference

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+            Const cSource = "Public Interface IC1 : Inherits IA1 : End Interface '' error
+Public Interface IC2 : Inherits IA2 : End Interface
+Public Interface IC3 : Inherits IB1 : End Interface
+Public Interface IC4 : Inherits IB2 : End Interface
+Public Interface IC5 : Inherits IB1, IA1 : End Interface '' error
+Public Interface IC6 : Inherits IB2, IA2 : End Interface
+Public Class C1 : Inherits B1 : End Class
+Public Class C2 : Inherits B2 : End Class
+"
+            Dim compc = CreateVisualBasicCompilation(code:=cSource, assemblyName:="C", referencedAssemblies:={ref}, referencedCompilations:={compa, compb})
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb]: Should also test with metadata references

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+            compc.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb]: Please use AssertTheseDiagnostics helper instead. This is the preferred way to verify diagnostics for VB tests.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: That would be a strange shape for the language specification. We'll suggest that.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb (hunk 27 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb]: I do not intend to test the English wording of the diagnostic.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: This is how all well-known attributes are matched even if the spec doesn't talk about that explicitly (ExtensionAttribute, etc.) I don't see any good reason to make an exception for this attribute.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb (hunk 27 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/InternalsImplementationOnlyTests.vb]: And I am asking you to do that and use baseline in a human readable form. The enum value isn't that interesting in my opinion.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: OK, done, but a helper was not necessary because the computation is just a method call.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb (hunk 20 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb]: Used `AttributeDescription` directly.
